File: frontend/lib/core/models/attendance.dart
class AttendanceModel {
  final int id;
  final int studentId;
  final int? teacherId;
  final int grade;
  final int period;
  final DateTime date;
  final String status; // 'present' | 'absent'
  final DateTime createdAt;

  const AttendanceModel({
    required this.id,
    required this.studentId,
    this.teacherId,
    required this.grade,
    required this.period,
    required this.date,
    required this.status,
    required this.createdAt,
  });

  bool get isPresent => status == 'present';

  factory AttendanceModel.fromJson(Map<String, dynamic> json) =>
      AttendanceModel(
        id: json['id'] as int,
        studentId: json['student_id'] as int,
        teacherId: json['teacher_id'] as int?,
        grade: json['grade'] as int,
        period: json['period'] as int,
        date: DateTime.parse(json['date'] as String),
        status: json['status'] as String,
        createdAt: DateTime.parse(json['created_at'] as String),
      );
}

class AttendanceSummaryModel {
  final int studentId;
  final int totalClasses;
  final int presentCount;
  final int absentCount;
  final double attendancePercentage;

  const AttendanceSummaryModel({
    required this.studentId,
    required this.totalClasses,
    required this.presentCount,
    required this.absentCount,
    required this.attendancePercentage,
  });

  factory AttendanceSummaryModel.fromJson(Map<String, dynamic> json) =>
      AttendanceSummaryModel(
        studentId: json['student_id'] as int,
        totalClasses: json['total_classes'] as int,
        presentCount: json['present_count'] as int,
        absentCount: json['absent_count'] as int,
        attendancePercentage:
            (json['attendance_percentage'] as num).toDouble(),
      );
}

File: backend/app/schemas/attendance.py
"""
Pydantic schemas for Attendance.
"""

from datetime import date, datetime
from typing import List, Optional

from pydantic import BaseModel, field_validator

from app.core.attendance_window import attendance_date_reason
from app.models.attendance import AttendanceStatus


class AttendanceCreate(BaseModel):
    student_id: int
    grade: int
    period: int
    date: date
    status: AttendanceStatus = AttendanceStatus.present


class AttendanceBulkCreate(BaseModel):
    """Used by teachers to mark attendance for a whole class in one request."""
    grade: int
    period: int
    date: date
    records: List[AttendanceCreate]

    @field_validator("date")
    @classmethod
    def _within_correction_window(cls, v: date) -> date:
        # Reject future dates and anything older than the backdate window, so a
        # teacher cannot mark attendance for a day that hasn't happened or
        # silently rewrite an old record. See core/attendance_window.py.
        reason = attendance_date_reason(v, date.today())
        if reason:
            raise ValueError(reason)
        return v


class AttendanceUpdate(BaseModel):
    status: AttendanceStatus


class AttendanceResponse(BaseModel):
    id: int
    student_id: int
    teacher_id: Optional[int] = None
    grade: int
    period: int
    date: date
    status: AttendanceStatus
    created_at: datetime

    model_config = {"from_attributes": True}


class AttendanceSummary(BaseModel):
    """Aggregated attendance stats for a student."""
    student_id: int
    total_classes: int
    present_count: int
    absent_count: int
    attendance_percentage: float
